src/inmemory_cache.rs

```diff
@@ -0,0 +1,134 @@
+    #[allow(dead_code)]
+    Filter(Arc<BloomFilter>),
+}
+
+#[derive(Clone, Copy)]
+pub enum CacheType {
+    Moka,
+}
+
+#[derive(Clone, Copy)]
+pub struct BlockCacheOptions {
+    pub max_capacity: u64,
+    pub cached_block_size: u32,
+    pub time_to_live: Option<Duration>,
+    pub time_to_idle: Option<Duration>,
+    pub cache_type: CacheType,
+}
+
+impl Default for BlockCacheOptions {
+    fn default() -> Self {
+        Self {
+            max_capacity: 64 * 1024 * 1024, // 64MB default max capacity
+            cached_block_size: 32,          // 32 bytes default
+            time_to_live: None,
+            time_to_idle: None,
+            cache_type: CacheType::Moka, // Default to Moka cache
+        }
+    }
+}
+
+#[async_trait]
+pub(crate) trait BlockCache: Send + Sync + 'static {
+    async fn get(&self, key: (SsTableId, usize)) -> CachedBlockOption;
+    async fn insert(&self, key: (SsTableId, usize), value: CachedBlock);
+    #[allow(dead_code)]
+    async fn remove(&self, key: (SsTableId, usize));
+    #[allow(dead_code)]
+    fn entry_count(&self) -> u64;
+}
+
+pub(crate) struct MokaCache {
+    inner: moka::future::Cache<(SsTableId, usize), CachedBlock>,
+}
+
+impl MokaCache {
+    pub fn new(options: BlockCacheOptions) -> Self {
+        let mut builder = moka::future::Cache::builder()
+            .weigher(move |_, _| options.cached_block_size)
+            .max_capacity(options.max_capacity);
+
+        if let Some(ttl) = options.time_to_live {
+            builder = builder.time_to_live(ttl);
+        }
+
+        if let Some(tti) = options.time_to_idle {
+            builder = builder.time_to_idle(tti);
+        }
+
+        let cache = builder.build();
+
+        Self { inner: cache }
+    }
+}
+
+#[async_trait]
+impl BlockCache for MokaCache {
+    async fn get(&self, key: (SsTableId, usize)) -> CachedBlockOption {
+        CachedBlockOption(self.inner.get(&key).await)
+    }
+
+    async fn insert(&self, key: (SsTableId, usize), value: CachedBlock) {
+        self.inner.insert(key, value).await;
+    }
+
+    async fn remove(&self, key: (SsTableId, usize)) {
+        self.inner.remove(&key).await;
+    }
+
+    fn entry_count(&self) -> u64 {
+        self.inner.entry_count()
+    }
+}
+
+/// Factory function to create the appropriate cache based on BlockCacheOptions
+pub(crate) fn create_block_cache(
+    options: Option<BlockCacheOptions>,
+) -> Option<Arc<dyn BlockCache>> {
+    if let Some(options) = options {
+        match options.cache_type {
+            CacheType::Moka => Some(Arc::new(MokaCache::new(options))),
+        }
+    } else {
+        None
+    }
+}
+
+/// wrapper around Option<CachedBlock> to provide helper functions
+pub(crate) struct CachedBlockOption(Option<CachedBlock>);
+
+impl CachedBlockOption {
+    pub(crate) fn block(&self) -> Option<Arc<Block>> {
+        match self {
+            CachedBlockOption(Some(CachedBlock::Block(block))) => Some(block.clone()),
+            _ => None,
```

Comment:
I kinda feel like these should raise a SlateDbError if `block` is called on a non-block struct.

Reply:
I am not sure raising an error is appropriate here since the impl is for an Option<CachedBlock>, so it kind of assumes it could be Option<Block>